File: modular_exponential/ModularExponential.java
import java.math.BigInteger;

public class ModularExponential {
    public static BigInteger modularExponential(long base, long power, long mod) {
        BigInteger result = BigInteger.ONE;
        base = base % mod;
        while(power > 0) {
            if(power % 2 == 1) {
                result = result.multiply(BigInteger.valueOf(base));
                result = result.mod(BigInteger.valueOf(mod));
            }
            power = power / 2;
            base = (base * base) % mod;
        }
        return result;
    }

    public static void main(String[] args) {
        long base = 78;
        long pow = 20;
        long mod = 65;
        System.out.println(modularExponential(base, pow, mod));
    }
}

File: modular_exponential/modular_exponential.go
package main

import "fmt"

// ModularExponential computes x^y
// Time Complexity : O(log(power))
func ModularExponential(base int64, power int, mod int64) int64 {
	result := int64(1)
	base = base % mod
	for power > 0 {
		if power%2 == 1 {
			result = (result * base) % mod
		}
		power = power >> 1
		base = (base * base) % mod
	}
	return result
}

func main() {
	base, power, mod := int64(2), 20, int64(100000)
	fmt.Println(ModularExponential(base, power, mod))
}
